Add more information to the documentation contribution guide

seankmartin opened the pull request:
# Description

Adds some more information around the different napari documentation resources and some more ways to build to the documentation contribution guide.

Main things:
1. Breaks down the different sources of napari docs and what they are written in (please let me know if I missed any!)
2. I've tried to highlight how you can contribute to the docs completely through GitHub if you like with no local setup. I think this would be great for simple changes and ease barrier to entry.
3. Mentioned the use of extra `make` commands such as `html-noplot` and `linkcheck-files`.
4. I also modified the makefile slightly to accept filenames to link check in the `linkcheck-files` command.

This is based on information I've picked up as I've been getting to grips with how to efficiently contribute to docs.
It is possible some details are slightly inaccurate as a result, so please let me know if anything I've written is a bit off!

I plan to make another PR at a later date with information about how to build the docs from Windows using git bash or WSL, but I think it best not to bloat this PR.

# References

Closes #34 - except for the comment "many docs changes will be fine to make in the GH interface or a plaintext editor, but a really nice feature of jupytext is that it allows you to run the .md files from a Jupyter Lab or Jupyter Notebook interface. This is one of the main reasons why we went with myst-md rather than plain notebooks. I think we should explicitly encourage this way of writing docs, including a final clear+run-all, as the preferred mode of contribution. Again, this could be done in a follow-up."

I've not used jupytext before, so I wouldn't feel too comfortable adding information about this. I'm not actually sure how running this document (for example) in Jupyter helps with writing it so I'm very open to input here! 😄

seankmartin committed:
First version of contribution cleanup

seankmartin committed:
Update to include npe2

seankmartin committed:
More details

seankmartin committed:
Attempt to clarify different doc sources and types

seankmartin committed:
Add information on link checks

melissawm commented on docs/developers/documentation/index.md:
```diff
@@ -6,27 +6,66 @@ documentation.
+## Contributing to the napari documentation without a local setup
+
+If you would like to see new documentation added to napari or want to see changes to existing documentation, but are not clear on these instructions or don't have the time to write it yourself, you can [open an issue](https://github.com/napari/docs/issues/new/choose).
+If you are adding new documentation or modifying existing documentation and would prefer a simpler workflow than the local setup guide described below,
+you can use the GitHub web interface to open your pull request
+by either [uploading file(s) from your computer](https://docs.github.com/en/repositories/working-with-files/managing-files/adding-a-file-to-a-repository),
+[creating and editing a new file](https://docs.github.com/en/repositories/working-with-files/managing-files/creating-new-files)
+or [editing an existing file](https://docs.github.com/en/repositories/working-with-files/managing-files/editing-files)
+on the [napari-docs](https://github.com/napari/docs) GitHub repository.
```
I love this, can we also add a note about doing this from your fork through a feature branch?

seankmartin replied:
Committed to add - "You can also first [fork](https://docs.github.com/en/get-started/quickstart/fork-a-repo) the [napari-docs](https://github.com/napari/docs) repository to your own GitHub account, create a [feature branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-and-deleting-branches-within-your-repository#creating-a-branch), upload/create/edit files through the GitHub web interface, and then [open a pull request from your fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork) back to [napari-docs](https://github.com/napari/docs)."

melissawm commented on docs/developers/documentation/index.md:
```diff
@@ -196,7 +237,7 @@ development installation of napari installed. [Examples](gallery)
 also be caught when previewing the built documentation. If your documentation
 change does not include code, you only need the napari docs dependencies
-installed.
+installed (in which case you could also preview your document with the [VScode Myst extension](https://marketplace.visualstudio.com/items?itemName=ExecutableBookProject.myst-highlight) or [Myst live preview](https://myst-parser.readthedocs.io/en/latest/live-preview.html#)).
```
I am actually not sure if this is useful. Because of how the docs are built, autogenerated images and formatting will not show up on the VScode interface and may give folks the false impression that something is broken.

seankmartin replied:
That's fair! I find it very useful to use the VScode Myst extension to live preview what I'm writing, with full understanding that I'm not looking at exactly what will be built due to some autogeneration. I understand that could be niche though. And I included the live preview link since that is similar to the Myst extension and I didn't want to vendor lock 😂 

I think there are two options then, open to either!
❌ Let's just remove any mention of Myst previews because the autogeneration differences will cause more confusion than the benefit of very quickly seeing document changes as you write.
✔️ Take that information out of brackets and expand it with explanation. Something like: "To see the markdown document structure and content change as you write it, you can use Myst markdown preview tools (such as Vscode and live preview). However, this preview page will have some differences to the final built html page, such as missing some autogenerated images and formatting.

psobolewskiPhD replied:
I like the 2nd option.

melissawm commented on docs/developers/documentation/index.md:
```diff
@@ -6,27 +6,66 @@ documentation.
+on the [napari-docs](https://github.com/napari/docs) GitHub repository.
+
+You will also be able to [preview the documentation](use-ci-artifacts) as it would appear on [napari.org](https://napari.org/dev) by
+downloading the built documentation via a link provided by a comment from the `github-actions` bot.
```
```suggestion
downloading the built documentation via a link provided by a comment from the `github-actions` bot, or previewing a live version of the website corresponding to your changes on [CircleCI](https://circleci.com/).
```

melissawm reviewed (commented):
Looks great, thanks @seankmartin - I left a couple suggestions below.

seankmartin committed:
Update docs/developers/documentation/index.md

Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

seankmartin committed:
Add info on using github web from a fork

seankmartin committed:
Fix small typo

seankmartin committed:
Change ssh clone to https clone

I think https will be easier for new users

psobolewskiPhD commented on docs/developers/documentation/index.md:
```diff
@@ -6,28 +6,69 @@ documentation.
+1. The main napari documentation is built from sources located at the
+[napari/docs](https://github.com/napari/docs) repository on GitHub.
+That repository is where all the narrative documentation (e.g. tutorials, how-to
+guides) pull requests should be made.
```
```suggestion
That repository is where all the narrative documentation (e.g. tutorials, how-to
guides, etc. on napari.org) pull requests should be made.
```

psobolewskiPhD commented on docs/developers/documentation/index.md:
```diff
@@ -6,28 +6,69 @@ documentation.
+[creating and editing a new file](https://docs.github.com/en/repositories/working-with-files/managing-files/creating-new-files)
+or [editing an existing file](https://docs.github.com/en/repositories/working-with-files/managing-files/editing-files)
+on the [napari-docs](https://github.com/napari/docs) GitHub repository.
+You can also first [fork](https://docs.github.com/en/get-started/quickstart/fork-a-repo) the [napari-docs](https://github.com/napari/docs) repository to your own GitHub account, create a [feature branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-and-deleting-branches-within-your-repository#creating-a-branch ...
```
```suggestion
It's best if you first [fork](https://docs.github.com/en/get-started/quickstart/fork-a-repo) the [napari-docs](https://github.com/napari/docs) repository to your own GitHub account, create a [feature branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-and-deleting-branches-within-your-repository#creating-a-branch), upload/create/edit files through the GitHub web interface, and then [open a pull request from your fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork) back to [napari-docs](https://github.com/napari/docs).
```

psobolewskiPhD commented on docs/developers/documentation/index.md:
```diff
@@ -6,28 +6,69 @@ documentation.
+You can also first [fork](https://docs.github.com/en/get-started/quickstart/fork-a-repo) the [napari-docs](https://github.com/napari/docs) repository to your own GitHub account, create a [feature branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-and-deleting-branches-within-your-repository#creating-a-branch ...
+
+You will also be able to [preview the documentation](use-ci-artifacts) as it would appear on [napari.org](https://napari.org/dev) by
+downloading the built documentation via a link provided by a comment from the `github-actions` bot, or previewing a live version of the website corresponding to your changes on [CircleCI](https://circleci.com/).
```
```suggestion
using the `Check the rendered docs here!` action at the bottom of your PR which will go to a preview site on [CircleCI](https://circleci.com/).
```

I don't think the bot is a thing anymore, just the CicleCi.

psobolewskiPhD replied:
It is possible to download the build artifact though...

seankmartin committed:
Replace napari with napari fork for consistency

psobolewskiPhD commented on docs/developers/documentation/index.md:
```diff
@@ -107,24 +148,26 @@ Run `jupytext your-notebook.ipynb --to myst` to create a new MyST version of you
 
-## 2. Update TOC
+(update-toc)=
+## 2. Update the table of contents
```
```suggestion
## 2. Update the table of contents (TOC)
```

psobolewskiPhD reviewed (approved):
I dig this. I made a few small comments.

Czaki commented on docs/developers/documentation/index.md:
````diff
@@ -6,28 +6,69 @@ documentation.
 ```bash
-git clone git@github.com:<your-username>/docs.git napari-docs
+git clone https://github.com/<your-username>/napari.git
+git clone https://github.com/<your-username>/docs.git napari-docs
````
Github does not allow for authorization via username and password. So it may be hard to push to repositories created in this way, as it may need to set up token authorization with not-so-good documentation and less googlable errors.

seankmartin replied:
Yeah that's a good point thanks @Czaki, I've gotten so used to the token authentication that I forgot this can be a bit awkward.

Do you think the SSH setup steps are clearer? I.e. changing that to
```
git clone git@github.com:<your-username>/napari.git
git clone git@github.com:<your-username>/docs.git napari-docs
```

Or would you recommend something else?

Czaki replied:
I prefer to check the current state of github documentation (I do not have time for this today) and do as they suggest and put link to their documentation somewhere. It will reduce the maintenance burden. 

Based on my memory, they suggest ssh.

seankmartin replied:
It doesn't seem to recommend any particular option from what I can see in the [Cloning a repository guide](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository). However, there is a helpful [Troubleshooting cloning guide](https://docs.github.com/en/repositories/creating-and-managing-repositories/troubleshooting-cloning-errors) which would probably good to to put in the docs.

psobolewskiPhD replied:
What about the `gh` command line tool? Is that a solution? it has some other upside like checking out PRs.

seankmartin committed:
Update docs/developers/documentation/index.md

Co-authored-by: Peter Sobolewski <76622105+psobolewskiPhD@users.noreply.github.com>

seankmartin committed:
Update docs/developers/documentation/index.md

Co-authored-by: Peter Sobolewski <76622105+psobolewskiPhD@users.noreply.github.com>

seankmartin committed:
Update docs/developers/documentation/index.md

Co-authored-by: Peter Sobolewski <76622105+psobolewskiPhD@users.noreply.github.com>

seankmartin committed:
Update docs/developers/documentation/index.md

Co-authored-by: Peter Sobolewski <76622105+psobolewskiPhD@users.noreply.github.com>

seankmartin committed:
Pull info about markdown live preview into tip

This was performed to make it clearer that the live preview is just to help
in your writing, and if you are not familiar with MyST/markdown
it can really help!
However, still need to build docs, not a replacement

seankmartin committed:
Update github clone options to use GitHub CLI

Also refer to the GitHub guides around cloning

seankmartin commented:
@Czaki, @psobolewskiPhD and @melissawm thank you all for the reviews!

On the two conversations we had:
1. [**MyST preview usefulness:**](https://github.com/napari/docs/pull/157/commits/4d5271e13d1e12d13ad59c8ec1c9c22137a40608) I've pulled the information about using a `MyST` preview into a tip block and mentioned that you still need to build locally / view the CI docs.
2. [**GitHub cloning difficulties:**](https://github.com/napari/docs/pull/157/commits/cbab64e49124f85c58629b7c7b370d0af3e85731) I've linked to the GitHub cloning guide and troubleshooting guide. Also recommended using the `GitHub CLI` as suggested by @psobolewskiPhD - I think it's the best option.

Happy to make changes on these!

melissawm commented on docs/developers/documentation/index.md:
```diff
@@ -6,37 +6,84 @@ documentation.
-[napari/docs](https://github.com/napari/docs) repository on GitHub. That
-repository is where all the narrative documentation (e.g. tutorials, how-to
-guides) pull requests should be made. Meanwhile, changes to docstrings or to the
+The napari documentation is built from multiple sources that are organised into repositories:
```
nit: we mostly use American spelling so we may need to decide on a style guide for this?

```suggestion
The napari documentation is built from multiple sources that are organized into repositories:
```

seankmartin replied:
Yeah a style guide sounds good on this!

melissawm commented on docs/developers/documentation/index.md:
```diff
@@ -6,37 +6,84 @@ documentation.
-[](add-examples)).
+[the guide on adding examples to the gallery](add-examples)).
+4. Changes to how [napari.org](https://napari.org/dev) appears site-wide are made to the
+[napari-sphinx-theme](https://github.com/napari/napari-sphinx-theme).
```
One thing that jumps out to me here is that we don't mention API documentation at all, but this is important. So maybe we need

```
5. The [API reference documentation](https://napari.org/dev/api/index.html) is autogenerated from the napari source code docstrings. Docstrings are written in the [reStructuredText format](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html) and any modifications to them should be submitted to the [napari/napari](https://github.com/napari/napari) repository.
```

seankmartin replied:
The API is briefly mentioned in point 3 as coming from the docstrings, but I like pulling that out into a separate point so will do, thanks.
I'll make point 3 about updating the examples gallery and pull the API mention into point 5 as you suggested

melissawm reviewed (approved):
Some nitpicks but otherwise looks good!

seankmartin committed:
Apply suggestions from code review

Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

seankmartin committed:
Add API docs information

seankmartin committed:
Merge branch 'main' into docs-contribution-cleanup

psobolewskiPhD committed:
Merge branch 'main' into docs-contribution-cleanup
